crates/runtime/src/dataconnector/github/rate_limit.rs

```diff
@@ -0,0 +1,233 @@
+    pub used: i32,
+    pub reset_time: DateTime<Utc>,
+    pub resource: String,
+}
+
+// See https://docs.github.com/en/graphql/overview/rate-limits-and-node-limits-for-the-graphql-api#checking-the-status-of-your-primary-rate-limit
+impl RateLimitInfo {
+    pub fn from_headers(headers: &HeaderMap) -> Option<Self> {
+        let limit = headers
+            .get("x-ratelimit-limit")?
+            .to_str()
+            .ok()?
+            .parse::<i32>()
+            .ok()?;
+        let remaining = headers
+            .get("x-ratelimit-remaining")?
+            .to_str()
+            .ok()?
+            .parse::<i32>()
+            .ok()?;
+        let used = headers
+            .get("x-ratelimit-used")?
+            .to_str()
+            .ok()?
+            .parse::<i32>()
+            .ok()?;
+        let reset = headers
+            .get("x-ratelimit-reset")?
+            .to_str()
+            .ok()?
+            .parse::<i64>()
+            .ok()?;
+        let resource = headers
+            .get("x-ratelimit-resource")?
+            .to_str()
+            .ok()?
+            .to_string();
+
+        let reset_time = Utc.timestamp_opt(reset, 0).unwrap();
+
+        Some(Self {
+            limit,
+            remaining,
+            used,
+            reset_time,
+            resource,
+        })
+    }
+}
+
+impl RateLimiter {
+    pub fn new() -> Self {
+        Self {
+            api_limit: Arc::new(RwLock::new(None)),
+            requests_made: 0,
+            start_time: SystemTime::now(),
+        }
+    }
+
+    pub async fn update_from_headers(&self, headers: &HeaderMap) {
+        if let Some(rate_limit) = RateLimitInfo::from_headers(headers) {
+            let mut api_limit = self.api_limit.write().await;
+            *api_limit = Some(rate_limit);
+        }
+    }
+
+    pub async fn check_rate_limit(&mut self) -> Result<(), String> {
+        // First check our internal hourly limit
+        if let Ok(elapsed) = self.start_time.elapsed() {
+            if elapsed >= Duration::from_secs(3600) {
+                // Reset counter if an hour has passed
+                self.requests_made = 0;
+                self.start_time = SystemTime::now();
+            } else if self.requests_made >= GITHUB_MAX_REQUESTS_PER_HOUR {
+                // Calculate time to wait until the hour is up
+                let wait_time = Duration::from_secs(3600) - elapsed;
+                warn!(
+                    "Internal rate limit reached ({}/{}). Waiting for {} seconds.",
+                    self.requests_made,
+                    GITHUB_MAX_REQUESTS_PER_HOUR,
+                    wait_time.as_secs()
+                );
+                tokio::time::sleep(wait_time).await;
+                self.requests_made = 0;
+                self.start_time = SystemTime::now();
+            }
+        }
+
+        // Then check GitHub's rate limit from API responses
+        if let Some(api_limit) = self.api_limit.read().await.clone() {
```

Comment:
Note: In the case of multiple API keys, there could be a situation where we rate-limit an API key that we should NOT.

Should the rate limit instance be per API key?

Reply:
Yes, they should be per API key. That should already be the case because this rate limit object is set per github client.